---
title: "mixMC"
date: "24 November 2016"
output:
  html_document:
    code_folding: show
    highlight: haddock
    df_print: kable

---
# mixMC: mixOmics for 16S Microbial Communities

**mixMC** [1](http://journals.plos.org/plosone/article/metrics?id=10.1371/journal.pone.0160169) is a multivariate framework for microbiome data analysis which takes into account the sparsity and compositionality of microbiome data. **mixMC** aims to identify specific associations between microbial communities and their type of habitat building on the hypothesis that multivariate methods can help identify microbial communities that modulate and influence biological systems as a whole. 

<br>

**mixMC** addresses the limitations of existing multivariate methods for microbiome studies and proposes unique analytical capabilities: it handles compositional and sparse data, repeated-measures experiments and multiclass problems; it highlights important discriminative features, and it provides interpretable graphical outputs to better understand the microbial communities contribution to each habitat.
<br>
<br>
<br>
<br>
<img src= "Images/MIXMCFramework.jpg" height="400px" width="500px" />
<br>
<br>

# To begin...

Load required libraries:

```{r opts_chunk$set(fig.width = 7, fig.height = 5), echo=TRUE, warning=FALSE, message=FALSE}
#install.packages("mixOmics")
library(mixOmics)
```


# Data
In the tabs under mixMC, examples are provided applying mixMC to microbiome data sets. To download the **full data sets**  and scripts see these links:

[Non-Repeated Measures - Koren mixMC example ](http://mixomics.org/wp-content/uploads/2017/01/Koren.zip)


[Repeated Measures - HMP 16s Data](http://mixomics.org/wp-content/uploads/2016/12/HMB.zip)

[Repeated Measures - Oral 16S Data](http://mixomics.org/wp-content/uploads/2016/12/Oral.zip)

# Microbiome data
Culture independent techniques, such as shotgun metagenomics and 16S rRNA amplicon sequencing have dramatically changed the way we can examine microbial communities. However, current statistical methods are limited in their scope to identifying and comparing bacteria driving changes in their ecosystem. This is partly due to the inherent properties of microbiome data.

<br>

The absence of microbial organisms from a large number of samples results in highly skewed count data with many zeros *sparse* count data. In addition, the varying sampling/sequencing depth between samples requires transformation of the count data into relative abundance (proportions) leading to *compositional* data. 

<br>

# mixMC: method improvements for microbiome 16S data analysis 
Compositional data pose statistical theoretical issues and potentially considerable misinterpretation with standard methods [2], as such data within a specimen sample sum to one, resulting in data residing in a simplex, rather than an Euclidian space. The solution proposed by several authors is to project the relative count data into an Euclidian space using log ratio transformations, such as centred log ratio transformation (CLR), before applying standard statistical techniques. The CLR transformation consists in dividing each sample by the geometric mean of its values and taking the logarithm [2],[3]. The transformation is symmetric, resulting in the retention of dimensions in the data [4]. 

<br>

**mixMC**, the mixOmics method sPLS-DA has been improved with CLR transformation and includes a multilevel decomposition for repeated measurements design that are commonly encountered in microbiome studies. The multilevel approach from [5] enables the detection of subtle differences when high inter-subject variability is present due to microbial sampling performed repeatedly on the same subjects but in multiple habitats. To account for subject variability the data variance is decomposed into  *within variation* (due to habitant) and *between subject variation* [6], similar to a within-subjects ANOVA in univariate analyses. 

<br>

Graphical outputs such as **plotLoadings()** can be used to represent the habitat in which the selected micro-organism is most present.
<br>
<br>
<br>
<br>

# References 
**1** [Lê Cao KA, Costello ME, Lakis VA, Bartolo F, Chua XY, et al. (2016) MixMC: A Multivariate Statistical Framework to Gain Insight into Microbial Communities. PLOS ONE 11(8): e0160169. doi: 10.1371/journal.pone.0160169](http://journals.plos.org/plosone/article/metrics?id=10.1371/journal.pone.0160169)

**2** [Aitchison, J., 1982. The statistical analysis of compositional data. Journal of the Royal Statistical Society. Series B (Methodological), pp.139-177.](https://scholar.google.com/scholar_url?url=http://www.jstor.org/stable/2345821&hl=en&sa=T&oi=gsb-gga&ct=res&cd=1&ei=bZ43WIeXNoLijAGxrrvwCw&scisig=AAGBfm29rId6wbzZnTe7T-ST4HCF0U4O4g)

**3** [Fernandes, A.D., Reid, J.N., Macklaim, J.M., McMurrough, T.A., Edgell, D.R. and Gloor, G.B., 2014. Unifying the analysis of high-throughput sequencing datasets: characterizing RNA-seq, 16S rRNA gene sequencing and selective growth experiments by compositional data analysis. Microbiome, 2(1), p.1.](https://scholar.google.com/scholar_url?url=https://microbiomejournal.biomedcentral.com/articles/10.1186/2049-2618-2-15&hl=en&sa=T&oi=gsb-gga&ct=res&cd=0&ei=gp43WM6sNIahjAHSwYSoAw&scisig=AAGBfm0PCTxgOShrYlSPE85FwQwuQVc7qw)

**4** [Filzmoser, P., Hron, K. and Reimann, C., 2009. Principal component analysis for compositional data with outliers. Environmetrics, 20(6), pp.621-632.](https://scholar.google.com/scholar_url?url=https://www.researchgate.net/profile/Peter_Filzmoser/publication/227528968_Principal_component_analysis_of_compositional_data_with_outliers/links/09e4151129a8c792bd000000.pdf&hl=en&sa=T&oi=gsb-gga&ct=res&cd=0&ei=qp43WMCHEouOjAGZ6Z-4Aw&scisig=AAGBfm1mB8frALUL699TNp-vVgk-XzGhOg)

**5** [Westerhuis, J.A., van Velzen, E.J., Hoefsloot, H.C. and Smilde, A.K., 2010. Multivariate paired data analysis: multilevel PLSDA versus OPLSDA. Metabolomics, 6(1), pp.119-128.](https://scholar.google.com/scholar_url?url=http://link.springer.com/article/10.1007/s11306-009-0185-z/fulltext.html&hl=en&sa=T&oi=gsb-gga&ct=res&cd=0&ei=xJ43WMGWNI6SjAHp3L6IAg&scisig=AAGBfm00SN6xghGwtNv5b5VZOxADOblHDQ)   

**6** [Liquet, B., Lê Cao, K.A., Hocini, H. and Thiébaut, R., 2012. A novel approach for biomarker selection and the integration of repeated measures experiments from two assays. BMC bioinformatics, 13(1), p.325.](https://scholar.google.com/scholar_url?url=http://www.biomedcentral.com/1471-2105/13/325&hl=en&sa=T&oi=gsb-gga&ct=res&cd=0&ei=4J43WPDdCMP82Aa6qLzQCQ&scisig=AAGBfm1OpbHVWALILoWtXKl7p-iQKRe5Mg)
