---
title: "mixMC Non-Multilevel KOREN DATASET"
date: "24 November 2016"
output:
  html_document:
    code_folding: show
    highlight: haddock
---
# Koren Data 16s

Here we apply mixMC to the 16S study of Koren et al. (2011) who examined the link between oral, gut and plaque microbial communities in patients with atherosclerosis and controls. Only healthy individuals were retained in the analysis. This study contained partially repeated measures from multiple sites including 15 unique patients samples from saliva and stool, and 13 unique patients only sampled from arterial plaque samples and we therefore considered a non multilevel analysis for that experimental design (i.e. no repeated measurements). The data were downloaded from the Qiita database. After pre-filtering, the data included 973 OTU for 43 samples.  Data were TSS normalized and log ratio transformations are applied in PCA and PLS-DA.

Load the latest version of [mixOmics](https://www.rdocumentation.org/link/%20www.mixOmics.org%2FmixMC?package=mixOmics&version=6.0.0).

```{r global_options, include=FALSE}
library(knitr)
knitr::opts_chunk$set(dpi = 100, echo= TRUE, warning=FALSE, message=FALSE)
```

```{r}
library(mixOmics)
```

# Data: Koren Data Set

Using the Koren data in this example we assume the data are prefiltered and normalised with TSS, as described in [mixMC normalisation](http://mixomics.org/mixmc/filtering-normalisation/).

```{r inputdata}
# normalised data
data("Koren.16S")
#ls(Koren.16S)
koren.TSS= Koren.16S$data.TSS
```

```{r saving files, echo = FALSE}
# create the following folder where the files will be saved 
path.output = 'Koren/'
```

We set up the outcome "Y" as a factor indicating the body sites of each sample. A repeated measures design is not nessessary as each sample is not repeated more then once. 

```{r, echo=TRUE, results= "hide"}
Y  =  Koren.16S$indiv$body_site
summary(Y)
```

# Unsupervised Analsysis with Principal Component Analysis

PCA enables visualisation of diversity patterns in an unsupervised analysis. We need to specify the log ratio transformation (choose between ‘CLR’ or ‘ILR’) and specify a sufficient number of components to understand how the variation in the data is explained per component.

```{r pca, fig.width= 4}
pca.res = pca(koren.TSS, ncomp = 10, logratio = 'CLR')
plot(pca.res)
```

The barplot above depicts the percentage of explained variance per component for a PCA with 10 components. It is a good indicator of how many components to retain in the final PCA model (here two components will be sufficient).

The sample plot below displays the samples colored according to Y, but note that PCA does not take Y into account in the analysis!

```{r plotIndiv PCA, fig.keep='all'}
#plot first component
plotIndiv(pca.res,  comp = c(1,2),
          pch = 16, ind.names = F, group = Y, col.per.group = color.mixo(1:3),
          legend = TRUE)
```

# Supervised Analysis and Selection of Discriminative OTUs with sPLS-DA

sPLS-DA performs variable (OTU) selection on each component, [see sPLS-DA](http://mixomics.org/methods/spls-da/) for details.

Firstly, to choose the number of components for sPLS_DA we run the function perf() for a **PLS-DA** model with no variable selection (often, ncomp = K-1 where K is the number of categories in the outcome Y is sufficient, but depends on the data). We choose 5-fold cross-validation repeated 10 times. To obtain a more reliable estimation of the error rate, the number of repeats should be increased (between 50 to 100). 

```{r plsda}
koren.plsda <- plsda(X = koren.TSS, Y, ncomp = nlevels(Y) - 1)

koren.perf.plsda <- perf(koren.plsda, validation = 'Mfold', folds = 5,
                    progressBar = FALSE, nrepeat = 10)

plot(koren.perf.plsda, overlay = 'measure', sd=TRUE)
```

Above is the plot of the classification error rate averaged across 5 folds and the 10 repeated CV for all prediction distances. BER stands for balanced error rate, which accounts for unbalanced number of samples per class. This step allows to choose the best prediction distance that will be input in the tune sPLS-DA step below.

```{r plot plsda}
#First two components
plotIndiv(koren.plsda , comp = c(1,2),
          group = Y, ind.names = FALSE, 
          ellipse = TRUE, legend = TRUE, title = 'KOREN-16S, PLSDA comp 1 - 2')
```

## Tuning sPLS-DA

Secondly, we tune keepX in sPLSDA with the chosen number of components + 1 (to double check our choice from the step above) using tune.splsda. We choose 5-fold cross-validation repeated 10 times in this example. To ensure a stable result we advise to set nrepeat = 50-100)

```{r tune splsda, eval = FALSE, message=FALSE}
# this chunk takes ~ 12 min to run
splsda.tune = tune.splsda(koren.TSS, 
                          Y = Y, 
                          ncomp = 3, 
                          multilevel = NULL, 
                          logratio = 'CLR',
                          test.keepX = c(seq(5,150, 5)), 
                          validation = c('Mfold'), 
                          folds = 5, 
                          dist = 'max.dist', # prediction distance can be chosen according to tune.plsda results
                          nrepeat = 10)
# may show some convergence issues for some of the cases, it is ok for tuning
```

```{r some splsda outputs, include=FALSE}
# To gain some computing time on the tuning, directly load the data
load('Koren/KOREN-tune-SPLSDA.RData')
```

Our tuning indicate the following optimal keepX to select on each component:

```{r}
splsda.tune$choice.keepX
```

The following command line will output the error rate according to the number of variable selected:

```{r}
kable(head(splsda.tune$error.rate))

# The plot will show the average error rate with respect to the keepX values tested:
plot(splsda.tune)
```

## sPLS-DA

We now run a sPLS-DA **non-multilevel** analysis. Note: with sPLS-DA we can only choose a CLR transformation (see details in our manuscript [here](http://mixomics.org/mixmc/multi-level-example/(https://scholar.google.com/scholar_url?url=http://biorxiv.org/content/biorxiv/early/2016/03/16/044206.full.pdf&hl=en&sa=T&oi=gsb-gga&ct=res&cd=0&ei=0dA7WPvyNIKvjAHq1LHYDw&scisig=AAGBfm04GmTLnXd5EizkZPC1ZlPH9IRsfA))). 


```{r}
# input parameters for sPLS-DA
# if you have not tuned the model,  you can set the optimal keepX to our results:

choice.keepX = c(150, 15, 145) # optimal keepX values according to the tuning criterion above.

choice.ncomp = length(choice.keepX) # the number of components


# the sPLS-DA
res.splsda = splsda(X = koren.TSS, 
                    Y = Y,
                    ncomp = choice.ncomp,
                    keepX = choice.keepX,
                    logratio= "CLR")
```

## sPLS-DA plots 

The sample plot below shows the sPLS-DA  first two components. The ellipse are 0.95 confidence interval ellipse for each body site.

```{r splsda-plot, echo= TRUE}
#for component 1 & 2
plotIndiv(res.splsda, 
          ind.names = F, 
          col.per.group = color.mixo(1:3), 
          comp = c(1,2),
          pch = 16, 
          ellipse = TRUE,
          legend = TRUE)
```

The code below outputs the first selected OTUs and their coefficient (from the loading vector) on the first component:

```{r,, echo=TRUE}
head(selectVar(res.splsda, comp = 1)$value)  # just a head
```

```{r, echo= FALSE}
# save the OTU selected on each component
list.splsda = list()
for(k in 1:choice.ncomp){
  list.splsda[[k]] = selectVar(res.splsda, comp = k)$name
}
```

## Evaluating sPLS-DA
The classification performance of the final sPLS-DA model can be assessed using the function *perf()*. The mean error rates per component and per prediction distance are output.The prediction distance can also be specified, see **?perf**. 

``` {r, inlcude = TRUE}
splsda.perf = perf(res.splsda, validation = 'Mfold', folds = 5, 
                   progressBar = FALSE, nrepeat = 10)

splsda.perf$error.rate

head(splsda.perf$error.rate.class)

plot(splsda.perf)
```

## Clustered Image maps

We represent clustered image maps (with Euclidian distance, Ward linkage set by default) for the OTU selected on each sPLS-DA component. The abundance values that are displayed are the normalised, log ratio transformed values. All OTUs selected by the sPLS-DA model are displayed, other options can include a specific component, see ?cim

```{r include= FALSE}
dev.off()
```

```{r echo=TRUE, fig.height= 8, fig.width=9}
cim(res.splsda, comp = 1, row.sideColors = color.mixo(Y))
```

## Contribution plots

The sPLS-DA selects the most discriminative OTUs that best characterize each body site indicated in Y. The contribution plots below displays the importance of each OTU in the sPLS-DA model and in which body site they are the most abundant (contrib = 'max'), according to the median (method = 'median'). Other options are available, see ?plotLoadings. 


```{r echo= TRUE}
#for component 1
plotLoadings(res.splsda, comp = 1, method = 'mean', contrib = 'max')
```

# References

1. [Koren, O., Knights, D., Gonzalez, A., Waldron, L., Segata, N., Knight, R., Huttenhower, C. and Ley, R.E., 2013. A guide to enterotypes across the human body: meta-analysis of microbial community structures in human microbiome datasets. PLoS Comput Biol, 9(1), p.e1002863](https://scholar.google.com/scholar_url?url=http://journals.plos.org/ploscompbiol/article%3Fid%3D10.1371/journal.pcbi.1002863&hl=en&sa=T&oi=gsb-gga&ct=res&cd=0&ei=bYQ_WKrEOsyT2AaftoLgDw&scisig=AAGBfm1kEjjIPN3hqJ9-wahg1glEXmYm_A)

2. [Lê Cao, K.-A., Boitard, S., Besse, P.: Sparse PLS Discriminant Analysis: biologically relevant feature selection and graphical displays for multiclass problems. BMC bioinformatics 12(1), 253 (2011)](https://scholar.google.com/scholar_url?url=http://bmcbioinformatics.biomedcentral.com/articles/10.1186/1471-2105-12-253&hl=en&sa=T&oi=gsb-gga&ct=res&cd=0&ei=y-g8WO2YMoKvjAHq1LHYDw&scisig=AAGBfm36JhKbmj5fZ9yykLKRPPqXYmyvPg)

3. [Lê Cao KA, Costello ME, Lakis VA, Bartolo F, Chua XY, et al. (2016) MixMC: A Multivariate Statistical Framework to Gain Insight into Microbial Communities. PLOS ONE 11(8): e0160169. doi: 10.1371/journal.pone.0160169](http://journals.plos.org/plosone/article/metrics?id=10.1371/journal.pone.0160169)

4. [Asnicar, F., Weingart, G., Tickle, T.L., Huttenhower, C. and Segata, N., 2015. Compact graphical representation of phylogenetic data and metadata with GraPhlAn. PeerJ, 3, p.e1029.](https://scholar.google.com/scholar_url?url=https://peerj.com/articles/1029/&hl=en&sa=T&oi=gsb-gga&ct=res&cd=0&ei=EtlAWM2XI46SjAGO2Jm4CQ&scisig=AAGBfm2IvCuL0Bjd50u1GMsL3UJhsF6JPQ)


