
 ---
title: "mixMC Multilevel HMP 16S Data"
output:
  pdf_document: default
  html_document:
    code_folding: show
    df_print: kable
    highlight: haddock
date: "22 November 2016"
---

# Human Microbiome Project 16s data 

Here we apply the mixMC framework on the Human Microbiome Most Diverse 16S data set, which includes OTU counts on the three most diverse bodysites: Subgingival plaque (Oral), Antecubital fossa (Skin) and Stool sampled from 54 unique individuals for a total of 162 samples.  The prefiltered dataset includes 1,674 OTU counts. The data were downloaded from the [Human Microbiome Project](http://hmpdacc.org/HMQCP/all/). The original data contained 43,146 OTU counts for 2,911 samples measured from 18 different body sites.

We focus here on the statistical exploration and analysis of the **repeated measurement** data set using multivariate projection-based approaches such as Principal Components Analysis (PCA) (data mining and exploration) and sparse Partial Least Squares Discriminant Analysis (sPLS-DA) (for identification of indicator species characterising each body site).

Launch mixOmics, which can be downloaded on our website. 
<span style="color:red">Dont forget to update to the [latest mixOmics version](http://mixomics.org/category/news/).</span>

```{r global_options, include=FALSE}
library(knitr)
knitr::opts_chunk$set(dpi = 100, echo= TRUE, warning=FALSE, message=FALSE, fig.align = 'center', 
                      fig.show=TRUE, fig.keep = 'all', out.width = '50%') 
```

```{r echo=TRUE, warning=TRUE, message=FALSE}
library(mixOmics)
```

# Data: Most Diverse Data Set

Using the NIH Human Microbial Project 16S data in this example we assume the data are prefiltered and normalised with TSS, as described in [mixMC normalisation](http://mixomics.org/mixmc/filtering-normalisation/).

```{r echo=TRUE}
data("diverse.16S")

# the 16S normalised data
data.mixMC = diverse.16S$data.TSS
```

We set up the outcome "Y" which is a factor indicating the body sites of each sample and "sample" which indicates the unique ID of each repeated individual (repeated measures design). 

```{r echo=TRUE, results="hide"}
# the outcome 
Y = diverse.16S$bodysite
# unique ID of each individual for multilevel analysis
sample = diverse.16S$sample
```

# Unsupervised Analsysis with PCA

Using unsupervised PCA allows for dimension reduction of the data and visualisation of diversity patterns in microbiome studies. Because we are using compositional data spurious results can arise if data is not transformed into Euclidean space. Therefore the data must be transformed using either [ILR](http://mixomics.org/mixmc/filtering-normalisation/) or [CLR](http://mixomics.org/mixmc/filtering-normalisation/). 

Here we chose the TSS normalisation ILR transformation as advised by Filmozer *et al* [1]. We prefer the ‘CLR’ log ratio transformation cas it is faster and can be used all throughout our framework (from PCA to sPLSDA), using the argument logratio = “CLR”).

```{r echo=TRUE}
pca.res = pca(data.mixMC, ncomp = 10, logratio = 'CLR', multilevel = sample)
#pca.res
plot(pca.res)
```

The PCA plot indicates that 2 components are sufficient to explain most of the variation in the data. 

**You can test different log ratio transformations or use different normalisation (CSS), and multilevel vs non multilevel to see the benefit of taking into account the repeated measures design in the PCA model.** Our team favours the normalisation TSS followed by CLR transformation.

```{r echo=TRUE}
plotIndiv(pca.res, 
          comp = c(1,2), # the components to plot
          pch = 16, 
          ind.names = F, 
          group = Y, 
          col.per.group = color.mixo(1:3),
          legend = TRUE)
```

We can observe a natural separation between those (very diverse) body sites. Next we run sPLSDA to identify which OTUs may drive this separation.

# Supervised Analysis and Selection of Discriminative OTUs with sPLS-DA

The mixMC frameworks uses the sPLS-DA multivariate analysis from mixOmics[3]. The sPLS-DA selects the most discriminative OTUs that best characterize each body site, [see sPLS-DA](http://mixomics.org/methods/spls-da/). 

Firstly, to choose the number of components for sPLS-DA we run the function perf() for a **PLS-DA** model with no variable selection (often, ncomp = K-1 where K is the number of categories in the outcome Y is sufficient, but depends on the data). We choose 5-fold cross-validation repeated 10 times. To obtain a more reliable estimation of the error rate, the number of repeats should be increased (between 50 to 100). 

Here we assess the performance of the PLSDA on `r nlevels(Y)` components.

```{r}
plsda <- plsda(X = data.mixMC, Y, ncomp = nlevels(Y), logratio = 'CLR', multilevel = sample)

perf.plsda <- perf(plsda, validation = 'Mfold', folds = 5,
                    progressBar = FALSE, nrepeat = 10)

plot(perf.plsda, overlay = 'measure', sd = TRUE)
```
The plot indicates a decrease in the classification error rate (i.e. an increase in classification performance) from one component to 2 components in the model. The BER stands for Balanced Error Rate and should be considered when we have an unbalanced number of samples per group. Here this is not the case so both overall and BER are overlapping. We can observe that the performance does not increase after 2 components, which suggest *ncomp = 2* for a final PLSDA model. Note that for the sparse PLSDA we may obtain a different ncomp.

For more details on the prediction distances, refer to Suppl info in Rohart et al. 2017 [4]. The distance 'max.dist' seem to lead to the best performance and will be kept for sPLSDA.


```{r}
plotIndiv(plsda , comp = c(1,2), ind.names = FALSE, 
          ellipse = TRUE, legend = TRUE, title = 'HMP Most Diverse, PLSDA comp 1 - 2')
```

Here we could check that the third component does not add additional information for the discrimination of the body sites.

```{r}
plotIndiv(plsda , comp = c(1,3), ind.names = FALSE, 
          ellipse = TRUE, legend = TRUE, title = 'HMP Most Diverse, PLSDA comp 1 - 3')
```



## Tuning sPLS-DA

The parameters to choose in sPLS-DA are the number of variables to select (keepX) and the number of components (ncomp) and to do this we used the function **tune.splsda()**. In this example the sPLS-DA tuning step is performed on 3 components, using 5-fold validation repeated 10 times.

The tune.splsda() needs to be performed prior to the sPLS-DA analysis to choose the parameters on a grid of keepX values (make sure you choose the appropriate M fold cross-validation and provide sufficient nrepeat = 50 in the evaluation model, except for ‘loo’ where it is run on 1 repeat), also check the stability of the features selected during the cross-validation process.

```{r tune splsda, eval = FALSE, echo = TRUE}
 # this chunk takes ~ 12 min to run
splsda.tune = tune.splsda(data.mixMC, Y, ncomp = 3, 
                           logratio = 'CLR',
                            multilevel = sample,
                           test.keepX = c(seq(5,150, 5)), validation = 'Mfold', 
                           folds = 5, dist = 'max.dist', nrepeat = 10)
# may show some convergence issues for some of the cases, it is ok for tuning
```

```{r echo=TRUE, results='hide'}
# To gain some computing time on the tuning, directly load the data
load('RData/mixMC-tune-sPLSDA.RData')
kable(head(splsda.tune$error.rate))
plot(splsda.tune)

# optimal number of variables to select on 3 comps:
select.keepX = splsda.tune$choice.keepX[1:3]

select.keepX

```

The graphic above shows that the error rate decreases when more components are included in sPLS-DA. The diamonds indicate the optimal keepX variables to select on each component based on the balanced error rate. The addtion of the third component seem to improve the perfomance and so our final sPLS-DA model will be run for 3 components.

## sPLS-DA

We now run a sPLS-DA multilevel analysis. **Note:** with sPLS-DA we can only choose a CLR transformation (see details [here](https://scholar.google.com/scholar_url?url=http://biorxiv.org/content/biorxiv/early/2016/03/16/044206.full.pdf&hl=en&sa=T&oi=gsb-gga&ct=res&cd=0&ei=0dA7WPvyNIKvjAHq1LHYDw&scisig=AAGBfm04GmTLnXd5EizkZPC1ZlPH9IRsfA).


```{r, echo=TRUE}
splsda.res <- splsda(data.mixMC, Y, ncomp = 3, logratio = 'CLR', multilevel = sample, keepX = select.keepX) 

plotIndiv(splsda.res, comp = c(1,2),
          ind.names = FALSE, 
          ellipse = TRUE, legend = TRUE,
          title = 'HMP Most Diverse, sPLSDA comp 1 - 2')
```

When we look however at the addition of the third component, it does not seem to discriminate the groups further. Hence we can focus primarily on the OTUs selected on the first 2 components.
```{r}
plotIndiv(splsda.res, comp = c(1,3),
          ind.names = FALSE, 
          ellipse = TRUE, legend = TRUE,
          title = 'HMP Most Diverse, sPLSDA comp 1 - 3')
```

## Evaluating sPLS-DA

The classification performance of the sPLS-DA multilevel model can be assessed using the function *perf()*. The mean error rates per component are output and type of distance are output. Here do not hesitate to increase the number of repeats for accurate estimations.

```{r Evaluating sPLS-DA, echo=TRUE}
# setting the seed so we get the same results with participants:
set.seed(34) 

splsda.perf = perf(splsda.res, validation = 'Mfold', folds = 5, 
                   progressBar = FALSE, nrepeat = 10, dist = 'max.dist')
splsda.perf$error.rate

plot(splsda.perf)
```

The performance plot confirms that 2 components are sufficient in the sPLS-DA model

## Contribution plots

The sPLS-DA selects the most discriminative OTUs that best characterize each body site. The contribution plots below display the abundance of each OTU (large abundance = large absolute value) and in which body site they are the most abundant for each sPLS-DA component. The contribution plots need to be interpreted in combination with the sample plot below to understand the similarities between body sites, in addition to 'which bacteria characterize those body sites?'

<br>
The code below outputs the first selected OTUs and their coefficient (from the loading vector) on the first component. Consider the absolute value as an indication of the importance of the OTU in the microbial signature, while the sign generally indicates positive / negative correlations between the OTUs TSS-CLR counts (see also with a plotVar).

```{r echo=TRUE, include= TRUE}
head(selectVar(splsda.res, comp = 1)$value) 
```

The plotLoadings shows that all OTUs selected on the first component are highly abundant in plaque.

```{r}
plotLoadings(splsda.res, comp = 1, method = 'mean', contrib = 'max',
             size.title = 1.0)

```

The plotLoadings shows that all OTUs selected on the second component are highly abundant in the two other bodysites. The sign indicate the opposition between the body sites.

```{r}
plotLoadings(splsda.res, comp = 2, method = 'mean', contrib = 'max',
             size.title = 1.0)

```

## Clustered Image Map 

```{r CIM, fig.height= 8, fig.width= 8}
#CIM plot
cim(splsda.res, row.sideColors = color.mixo(Y))
```

# References

1. [Filzmoser, P., Hron, K., Reimann, C.: Principal component analysis for compositional data with outliers. Environmetrics 20(6), 621–632 (2009)](https://scholar.google.com/scholar_url?url=https://www.researchgate.net/profile/Peter_Filzmoser/publication/227528968_Principal_component_analysis_of_compositional_data_with_outliers/links/09e4151129a8c792bd000000.pdf&hl=en&sa=T&oi=gsb-gga&ct=res&cd=0&ei=4ZA7WO-bGMyT2AaosKqwDQ&scisig=AAGBfm1mB8frALUL699TNp-vVgk-XzGhOg)

2. [Le Cao KA, Costello ME, Lakis VA, Bartolo F, Chua XY, et al. (2016) MixMC: A Multivariate Statistical Framework to Gain Insight into Microbial Communities. PLOS ONE 11(8): e0160169. doi: 10.1371/journal.pone.0160169](http://journals.plos.org/plosone/article/metrics?id=10.1371/journal.pone.0160169)


3. [Le Cao, K.A., Boitard, S. and Besse, P., 2011. Sparse PLS discriminant analysis: biologically relevant feature selection and graphical displays for multiclass problems. BMC bioinformatics, 12(1), p.1.](https://scholar.google.com/scholar_url?url=http://bmcbioinformatics.biomedcentral.com/articles/10.1186/1471-2105-12-253&hl=en&sa=T&oi=gsb-gga&ct=res&cd=0&ei=7dA7WJCXNI6SjAHp3L6IAg&scisig=AAGBfm36JhKbmj5fZ9yykLKRPPqXYmyvPg)

4. [Rohart F, Gautier B, Singh A, Lê Cao K-A (2017). mixOmics: an R package for 'omics feature selection and multiple data integration.](http://mixomics.org/a-propos/publications/)

